Preserve comments, don't silently truncate config files

quinox opened the pull request:
Changes the nginx parser in three ways:
* Partly fixes #450 (it doesn't solve the preservation of newlines). 
* Fixes the issue where your nginx configuration file could get truncated on parsing errors; now instead it will simply abort.
* Always end nginx configuration files with a newline character

The fix for #450 has a limitation: inline comments are written out as two separate lines, see `minimalistic_comments.conf`/`minimalistic_comments.new.conf`. It looks like fixing this will be tricky.

Running `tox` passes everything except `py26` due to a lack of Python 2.6 on my machine.

diracdeltas commented:
At a glance, this looks awesome. I'll try to review this weekend. Thanks!

diracdeltas commented on letsencrypt-nginx/letsencrypt_nginx/tests/testdata/etc_nginx/minimalistic_comments.conf:
```diff
@@ -0,0 +1,2 @@
+# Use bar.conf when it's a full moon!
+include foo.conf;  # Kilroy was here
```
Please add comments inside a block to this test case. Ex:
```
events {
    # some events
    worker_connections 1024;
    # another comment
}
```

diracdeltas commented on letsencrypt-nginx/letsencrypt_nginx/tests/nginxparser_test.py:
```diff
@@ -104,8 +106,17 @@ def test_parse_from_file2(self):
+        try:
+            load(open(util.get_data_filename('broken.conf')))
+        except ParseException as ex:
+            caught = ex
```
I think you should be able to use `self.assertRaises` here.

diracdeltas commented:
`tox` is broken for me on OSX inside a virtualenv right now, so I haven't actually run this. But it looks good other than the minor comments above. Thanks!

kuba commented on letsencrypt-nginx/letsencrypt_nginx/tests/util.py:
```diff
@@ -59,3 +59,18 @@ def get_nginx_configurator(
+
+
+def filter_comments(tree):
+    '''Filter comment nodes from parsed configurations.'''
```
nit: pls use `"""` instead of `'''`

kuba commented on letsencrypt-nginx/letsencrypt_nginx/tests/nginxparser_test.py:
```diff
@@ -120,9 +131,25 @@ def test_dump_as_file(self):
+    def test_comments(self):
+        parsed = load(open(util.get_data_filename('minimalistic_comments.conf')))
+
+        _file = open(util.get_data_filename('minimalistic_comments.new.conf'), 'w')
```
please use `with open...` context manager

kuba commented on letsencrypt-nginx/letsencrypt_nginx/tests/nginxparser_test.py:
```diff
@@ -120,9 +131,25 @@ def test_dump_as_file(self):
         self.assertEquals(parsed, parsed_new)
 
+    def test_comments(self):
+        parsed = load(open(util.get_data_filename('minimalistic_comments.conf')))
```
fd not closed, this should be wrapped in context manager

kuba commented on letsencrypt-nginx/letsencrypt_nginx/tests/nginxparser_test.py:
```diff
@@ -104,8 +106,17 @@ def test_parse_from_file2(self):
                  ['try_files', '$uri @rewrites']]]]]])
 
+    def test_abort_on_parse_failure(self):
+        # When we drop support for 2.6 we can use the context manager here
```
2.6 supports context managers

quinox replied:
Yes, but unfortunately `assertRaises` can't used as a context manager until 2.7.

kuba replied:
That's true.

kuba commented on letsencrypt-nginx/letsencrypt_nginx/tests/nginxparser_test.py:
```diff
@@ -59,10 +61,10 @@ def test_dump_as_string(self):
 
     def test_parse_from_file(self):
-        parsed = load(open(util.get_data_filename('foo.conf')))
+        parsed = util.filter_comments(load(open(util.get_data_filename('foo.conf'))))
```
this should be wrapped in `with open...`

kuba commented on letsencrypt-nginx/letsencrypt_nginx/tests/nginxparser_test.py:
```diff
@@ -85,7 +87,7 @@ def test_parse_from_file(self):
 
     def test_parse_from_file2(self):
-        parsed = load(open(util.get_data_filename('edge_cases.conf')))
+        parsed = util.filter_comments(load(open(util.get_data_filename('edge_cases.conf'))))
```
same as above

quinox commented:
Hi diracdeltas and kuba,

Thanks for the feedback, I've incorporated it all and rebased.

I've extended the comment testcase with a comment inside a block, I did not realize these lines were processed differently from top-level lines. Actually, with this PR that's no longer the case: I just refactored the dumper code a little bit so top-level lines are processed the same as nested lines, simplifying the code and reducing the LOC.

When I did this I noticed a bug of the parser where parameter-less statements only worked on block-level, not on top-level. I fixed this by adding ```default=None``` to the optional parameter group.

I also realized there was this weird `yield spacer` statement; `yield ''` makes sense, `yield spacer * current_indent` makes kind of sense but simply outputting a single space no matter how many blocks deep you are looked odd to me. I've replaced it with an empty line now.

I've added context managers to the testcases and the docstrings are written with triple double quotes.

kuba commented:
I don't know much about nginx plugin, but this generally looks good and works on my jessie test vps. :+1:

jdkasten commented:
Awesome work @quinox!  Thanks for the review @diracdeltas and @kuba.  -- and thanks for testing it out.

diracdeltas commented:
Thanks again @quinox. If you have the time, these would be good fixes to contribute back upstream to nginxparser (https://github.com/fatiherikli/nginxparser) which the letsencrypt module is based on.
